Added XML declaration check & `Source#skip_spaces` method

naitoh opened the pull request:
## Why?

### Added XML declaration check

- The version attribute is required in XML declaration.
- Only version attribute, encoding attribute, and standalone attribute are allowed in XML declaration.
- XML declaration is only allowed once.

See: https://www.w3.org/TR/xml/#NT-XMLDecl

### Added `Source#skip_spaces` method

In the case of `@source.match?(/\s+/um, true)`, if there are no spaces at the beginning, I want to stop reading immediately.
However, it continues to read the buffer until it finds a match, but it never finds a match.
As a result, it continues reading until the end of the file.

In the case of large XML files, drop_parsed_content occur frequently until the buffer is cleared, which may affect performance.


## Benchmark

```
                         before       after  before(YJIT)  after(YJIT) 
                 dom     32.534      35.130        54.559       53.528 i/s -     100.000 times in 3.073715s 2.846540s 1.832883s 1.868189s
                 sax     44.785      44.089        78.303       77.842 i/s -     100.000 times in 2.232907s 2.268138s 1.277093s 1.284657s
                pull     51.750      51.105        90.819       90.658 i/s -     100.000 times in 1.932351s 1.956759s 1.101094s 1.103050s
              stream     51.427      51.444        89.820       88.971 i/s -     100.000 times in 1.944502s 1.943855s 1.113340s 1.123960s

Comparison:
                              dom
        before(YJIT):        54.6 i/s 
         after(YJIT):        53.5 i/s - 1.02x  slower
               after:        35.1 i/s - 1.55x  slower
              before:        32.5 i/s - 1.68x  slower

                              sax
        before(YJIT):        78.3 i/s 
         after(YJIT):        77.8 i/s - 1.01x  slower
              before:        44.8 i/s - 1.75x  slower
               after:        44.1 i/s - 1.78x  slower

                             pull
        before(YJIT):        90.8 i/s 
         after(YJIT):        90.7 i/s - 1.00x  slower
              before:        51.8 i/s - 1.75x  slower
               after:        51.1 i/s - 1.78x  slower

                           stream
        before(YJIT):        89.8 i/s 
         after(YJIT):        89.0 i/s - 1.01x  slower
               after:        51.4 i/s - 1.75x  slower
              before:        51.4 i/s - 1.75x  slower
```

- YJIT=ON : 0.98x - 1.00x faster
- YJIT=OFF : 0.98x - 1.07x faster

kou commented:
> ### IOSource#match? read_keep param added
> 
> In the case of `@source.match?(/\s+/um, true)`, if there are no spaces at the beginning, I want to stop reading immediately. However, it continues to read the buffer until it finds a match, but it never finds a match. As a result, it continues reading until the end of the file.
> 
> In the case of large XML files, drop_parsed_content occur frequently until the buffer is cleared, which may affect performance.

How about adding `Source#skip_spaces` or something instead of adding a new argument to `Source#match?`?
I feel that adding `read_keep` parameter reduces readability...

naitoh committed:
Add `Source#skip_spaces` method

## Why?
In the case of `@source.match?(/\s+/um, true)`, if there are no spaces at the beginning, I want to stop reading immediately.
However, it continues to read the buffer until it finds a match, but it never finds a match.
As a result, it continues reading until the end of the file.

In the case of large XML files, drop_parsed_content occur frequently until the buffer is cleared, which may affect performance.

naitoh commented:
I modified it to use `Source#skip_spaces?`.

kou commented on lib/rexml/parsers/baseparser.rb:
```diff
@@ -787,6 +844,25 @@ def scan_quote
         end
       end
 
+      def get_attribute(name)
```
How about `parse_attribute_value_with_equal` or something?

naitoh replied:
OK.
I see.

naitoh commented on lib/rexml/source.rb:
```diff
@@ -142,14 +143,18 @@ def match(pattern, cons=false)
     end
 
-    def match?(pattern, cons=false)
+    def match?(pattern, cons=false, read_keep=true)
```
Oh, sorry.

naitoh commented on lib/rexml/source.rb:
```diff
@@ -142,14 +143,18 @@ def match(pattern, cons=false)
       end
     end
 
+    def skip_spaces?
```
OK.
I see.

naitoh commented on lib/rexml/parsers/baseparser.rb:
```diff
@@ -735,37 +736,93 @@ def process_comment
+          raise ParseException.new("Malformed XML: XML declaration is not at the start", @source)
+        end
+        unless @source.skip_spaces?
+          raise ParseException.new("Malformed XML: XML declaration is missing spaces", @source)
```
OK.
I see.

naitoh commented on lib/rexml/parsers/baseparser.rb:
```diff
@@ -290,7 +293,7 @@ def pull_event
               end
               name = parse_name(base_error_message)
-              @source.match?(/\s*/um, true) # skip spaces
+              @source.match?(Private::CUT_SPACES, true)
```
Oh, it is cool!
Thanks!

naitoh committed:
Added XML declaration check

## Why?

- The version attribute is required in XML declaration.
- Only version attribute, encoding attribute, and standalone attribute are allowed in XML declaration.
- XML declaration is only allowed once.

See: https://www.w3.org/TR/xml/#NT-XMLDecl

tompng commented on lib/rexml/source.rb:
```diff
@@ -65,9 +65,10 @@ class Source
     attr_reader :encoding
 
     module Private
+      SKIP_SPACES = /\s+/um
```
About naming. How about `SPACES` `SPACES_PATTERN`?
I think most regexp constant names in REXML represents what it is or what it matches, but unlike other constants, the naming of `SKIP_SPACES` and `CUT_SPACES` focuses on action.

If it is only used in `def skip_spaces`, simply `@scanner.skip(/\s+/um)` might be good enough.

naitoh replied:
Changed to `SPACES_PATTERN`.
Thank you!

kou reviewed (approved):
+1

naitoh commented:
Thank you for your reviews.
